### ISSUE TITLE
ingester: investigate high ttbr in low-traffic environments

### ISSUE THREAD

**Issue by lukebond** [influxdata/influxdb_iox#5828]
in our staging environments we're seeing high TTBR values, of 3+ days, when it appears to be functioning normally. there may be an issue on startup seeking the write buffer to the correct point and it's taking an old op from kafka. in an environment with even a moderate rate of writes this wouldn't be noticeable in TTBR, but in an environment with little-to-no traffic even an off-by-one error could report very high TTBR if nothing was written for days.

**Comment by lukebond** [comment]
i turned on trace logging for the ingester stream handler in staging and followed logs for the highest TTBR shard.

```
2022-10-11 12:13:53 level=info msg="Creating new partition-specific broker connection" topic=iox-shared partition=51 target="rskafka::client::partition" location="/usr/local/cargo/git/checkouts/rskafka-8a50f1e93cf86eb0/3208e47/src/client/partition.rs:368"
2022-10-11 12:13:53 level=info msg="Detected leader" topic=iox-shared partition=51 leader=1 target="rskafka::client::partition" location="/usr/local/cargo/git/checkouts/rskafka-8a50f1e93cf86eb0/3208e47/src/client/partition.rs:346"
2022-10-11 12:13:53 level=info msg="Detected leader" topic=iox-shared partition=51 leader=1 target="rskafka::client::partition" location="/usr/local/cargo/git/checkouts/rskafka-8a50f1e93cf86eb0/3208e47/src/client/partition.rs:346"
2022-10-11 12:13:53 level=info msg="Created new partition-specific broker connection" topic=iox-shared partition=51 leader=1 target="rskafka::client::partition" location="/usr/local/cargo/git/checkouts/rskafka-8a50f1e93cf86eb0/3208e47/src/client/partition.rs:433"
2022-10-11 12:13:53 level=info msg="Seek stream" shard_index=51 min_unpersisted_sequence_number=13 target="ingester::handler" location="ingester/src/handler.rs:238"
2022-10-11 12:13:53 level=trace msg="decoded dml operation" kafka_topic=iox-shared shard_index=51 op_size=390455 op_namespace=833f7e3e374cfebb_31db760575c68805 op_sequence_number=Some(SequenceNumber(13)) target="ingester::stream_handler::handler" location="ingester/src/stream_handler/handler.rs:358"
2022-10-11 12:13:53 level=trace msg="successfully applied dml operation" kafka_topic=iox-shared shard_index=51 should_pause=false target="ingester::stream_handler::handler" location="ingester/src/stream_handler/handler.rs:375"
2022-10-11 12:13:53 level=trace msg="reporting TTBR for shard (ms)" kafka_topic=iox-shared shard_index=51 delta=340983945 target="ingester::stream_handler::handler" location="ingester/src/stream_handler/handler.rs:399"
2022-10-11 13:13:53 level=info msg="partition is over age threshold, persisting" shard_id=51 partition_id=166323 first_write=2022-10-11T12:13:53.544909969+00:00 last_write=2022-10-11T13:13:51.424258966+00:00 bytes_written=4656960 rows_written=8640 first_sequence_number=SequenceNumber(182270) age=3600.15458355s target="ingester::lifecycle" location="ingester/src/lifecycle.rs:334"
```

what we can learn from this:
- it fetches the min unpersisted sequence number from the catalog (which is `13`)
- it seeks to that point on the kafka client
- it receives operation with sequence number 13 from the stream
- it successfully applies the operation
- it reports the TTBR of 3+ days, which is presumably correct
- (up until this point, this is all good)
- the odd bit is the last line: when persisting that partition it says the minimum sequence number is `182270` and not `13`. for some reason that op with sequence number `13` isn't persisted :eyes:

**Comment by domodwyer** [comment]
@lukebond this is great so far!

Regarding your observation of:

```
2022-10-11 13:13:53 level=info msg="partition is over age threshold, persisting" shard_id=51 partition_id=166323 first_write=2022-10-11T12:13:53.544909969+00:00 last_write=2022-10-11T13:13:51.424258966+00:00 bytes_written=4656960 rows_written=8640 first_sequence_number=SequenceNumber(182270) age=3600.15458355s target="ingester::lifecycle" location="ingester/src/lifecycle.rs:334"
```

Which looks suspicious because as you point out, 13 != 182270 for `first_sequence_number`! However we don't know that the partition being persisted in that log line is the one that received a write with `op=13`. A shard (as you likely know already) contains many tables, each with one or more partitions, and those partitions can be persisted **out of sequence order** within a shard. An example being (all in one shard):

* Write comes in with `sequence_number=1`, wrote to `partition=1`
* Write comes in with `sequence_number=2`, wrote to `partition=2`
* A huge write comes in with `sequence_number=3`, wrote to `partition=2` again - this pushes it over the "number of rows allowed in a partition" which makes it eligible for persisting.

At this point, a shard has the following ops with sequence numbers 1-3.

```text
            ┌─────────────┬─────────────┬─────────────┐
            │    Write    │    Write    │ Large Write │
            │    op=1     │    op=2     │    op=3     │
            └─────────────┴─────────────┴─────────────┘
                   ▲
                   │      ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
                   │                    │
            
              Partition 1          Partition 2
```

Once a minute a task runs that goes and grabs all the [partitions that are eligible][maybe_persist] (which introduces some non-deterministic / timing based randomness) and persists them. In the example above, partition 1 has lower sequence numbers, but partition 2 is big enough that it needs persisting, so it you would see a log line like:

```
msg="partition is over age threshold, persisting" shard_id=X partition_id=2 first_sequence_number=SequenceNumber(2)
```

And partition 2 (and the later writes 2 & 3) would be persisted before partition 1 (which had write 1).

We need to find out what partition ID that write is applied to (perhaps more `trace` logging needs adding to help with that) and then see if that partition is ever persisted. If we see a persist line for it, then the problem will be related to storing the resume point / loading the resume point / correctly skipping to the resume point / ignoring already applied ops during skipping. If we never see a persist for that partition, it will be the persist eligibility code that is at fault :+1:

Given all partitions have the same age timeout, I'd expect all the partitions that received writes at startup to be persisted at the same time (if they haven't already) so my money is on the latter!

[maybe_persist]:https://github.com/influxdata/influxdb_iox/blob/b5973b557515e77839ee5462a69416c372d6e42a/ingester/src/lifecycle.rs#L337-L435

**Comment by lukebond** [comment]
i ran another test, gathering logs for a period just over the old/cold threshold, and followed the logs for shard `40` on ingester 1, which is one of the highest TTBR shards. some findings below:

```
$ egrep 'shard_index=40|shard_id=40' ingester-1-logs.txt
2022-10-13 10:39:18 level=info msg="Seek stream" shard_index=40 min_unpersisted_sequence_number=13 target="ingester::handler" location="ingester/src/handler.rs:238"
2022-10-13 10:39:18 level=trace msg="decoded dml operation" kafka_topic=iox-shared shard_index=40 op_size=390455 op_namespace=833f7e3e374cfebb_31db760575c68805 op_sequence_number=Some(SequenceNumber(13)) target="ingester::stream_handler::handler" location="ingester/src/stream_handler/handler.rs:358"
2022-10-13 10:39:18 level=trace msg="successfully applied dml operation" kafka_topic=iox-shared shard_index=40 should_pause=false target="ingester::stream_handler::handler" location="ingester/src/stream_handler/handler.rs:375"
2022-10-13 10:39:18 level=trace msg="reporting TTBR for shard (ms)" kafka_topic=iox-shared shard_index=40 delta=247433027 target="ingester::stream_handler::handler" location="ingester/src/stream_handler/handler.rs:399"
```

this shows the same results as above, i.e. none of the extra trace logs in the buffering/writing chain of function calls is being logged for this partition.

i [added a trace log](https://github.com/influxdata/influxdb_iox/blob/main/ingester/src/data/table.rs#L163-L167) to detect skipped writes that have already been persisted, but that's not happening for shard 40:

```
$ grep 'skipping already-persisted write' ingester-1-logs.txt | grep 'shard_id=40'
```

(i.e. no output) so we know it's not being skipped.

i [added a trace log](https://github.com/influxdata/influxdb_iox/blob/main/ingester/src/data/partition.rs#L251-L255) for when a write is buffered, printing the min/max sequence number but since it doesn't state the shard ID i can't correlate them. however sequence number 13 is not in the output:

```
$ grep 'buffered write' ingester-1-logs.txt | grep 'SequenceNumber(13)'
```

that doesn't mean there isn't a `buffered write` printout related to shard 40 but i can't tell. another trace log on the caller to that function would allow us to correlate the passed-in sequence number to the shard id.

i [added a trace log](https://github.com/influxdata/influxdb_iox/blob/main/ingester/src/lifecycle.rs#L100-L110) for when the buffering is completed, but our shard ID or seq number don't appear:

```
$ grep 'logged write' ingester-1-logs.txt | grep 'shard_id=40'
$ grep 'logged write' ingester-1-logs.txt | grep 'SequenceNumber(13)'
```

i also [added a trace log](https://github.com/influxdata/influxdb_iox/blob/main/ingester/src/lifecycle.rs#L553-L558) for when the min_unpersisted_sequence_number value gets updated, but it isn't printing for this shard:

```
$ grep 'updated min_unpersisted_sequence_number' ingester-1-logs.txt | grep 'shard_id=40'
```

(i.e. nothing) this is not surprising as we know this value (`13`) is the same across pod restarts.

finally, and this isn't using any new trace logs, but we can see that partition 40 isn't getting persisted at all:

```
$ grep 'persisting' ingester-1-logs.txt | grep 'shard_id=40'
```

(i.e. nothing)

so.... what have we learned?

- we resume at the correct sequence number from the catalog, as we learned in the earlier logs i grabbed, above
- we pull the write off kafka with that sequence number and apply it
- we are unable to be sure from the traces we have whether or not the write was buffered but it looks like it doesn't (more on this below)
- the check to see if a write should be skipped because it has already been written didn't fire for this op; it wasn't skipped
- we never received a persist message for shard 40
- despite the op being passed to `.apply()` (we know this because it happens a few lines after the `decoded dml operation` trace line and there is no chance for it to return before, see [here](https://github.com/influxdata/influxdb_iox/blob/main/ingester/src/stream_handler/handler.rs#L358-L373), and we see the log that it was successfully applied) we didn't see a `logged write` message for it. this is surprising to me. if something caused it to not be buffered then that may be the issue, and the reason why it isn't picked up as a persist candidate. looking at the code i can see no way that `.apply()` wouldn't result in a `logged write` message without there being an error :woman_shrugging:

**Comment by lukebond** [comment]
so it turns out `shard_id` and `shard_index` are different things :woman_facepalming: 

this should help: https://github.com/influxdata/influxdb_iox/pull/5851 then i'll gather logs again

**Comment by lukebond** [comment]
fixed by #5854
